src/core/models.py

```diff
@@ -125,6 +132,18 @@ def __init__(
         self._balancer_uris = balancer_uris
         self._controller_password = controller_password
 
+    def _fetch_from_secrets(self, group, field):
+        if not self.relation:
+            return ""
+
+        if secrets_uri := self.relation.data[self.relation.app].get(
+            f"{PROV_SECRET_PREFIX}{group}"
+        ):
+            if secret := self.data_interface._model.get_secret(id=secrets_uri):
+                return secret.get_content().get(field, "")
+
+        return ""
```

Comment:
**praise:** This is much better 👍🏾  - thank you for the improvement!

Reply:
Honestly, I'm a little afraid to do so given the nature of `_fetch_relation_data_with_secrets` which has a bit of legacy :(